src/zarr/core/type_check.py

```diff
@@ -0,0 +1,505 @@
+        # Pass literal arguments through as-is, they are values, not types to resolve.
+        return Literal.__getitem__(args)
+
+    # Handle types.UnionType (Python 3.10+ union syntax like str | int)
+    if isinstance(tp, types.UnionType):
+        # Don't try to reconstruct UnionType, convert to typing.Union
+        resolved_args = tuple(_resolve_type(a, type_map, globalns, localns, _seen) for a in args)
+        return typing.Union[resolved_args]  # noqa: UP007
+
+    # Recurse into other generics
+    if origin and args:
+        new_args = tuple(_resolve_type(a, type_map, globalns, localns, _seen) for a in args)
+        # Special handling for single-argument generics like NotRequired, ReadOnly
+        if len(new_args) == 1:
+            return origin[new_args[0]]  # Pass single argument, not tuple
+        else:
+            return origin[new_args]  # Pass tuple for multi-argument generics
+
+    return tp
+
+
+def check_type(obj: Any, expected_type: Any, path: str = "value") -> TypeCheckResult:
+    """
+    Check if `obj` is of type `expected_type`.
+    """
+    origin = get_origin(expected_type)
+
+    if origin in (NotRequired, ReadOnly):
+        args = get_args(expected_type)
+        inner_type = args[0] if args else Any
+        return check_type(obj, inner_type, path)
+
+    if expected_type is Any:
+        return TypeCheckResult(True, [])
+
+    if origin is typing.Union or isinstance(expected_type, types.UnionType):
+        return check_union(obj, expected_type, path)
+
+    if origin is typing.Literal:
+        return check_literal(obj, expected_type, path)
+
+    if expected_type is None or expected_type is type(None):
+        return check_none(obj, path)
+
+    # Check for TypedDict (now unified)
+    if (origin and _is_typeddict_class(origin)) or _is_typeddict_class(expected_type):
+        return check_typeddict(obj, expected_type, path)
+
+    if origin is tuple:
+        return check_tuple(obj, expected_type, path)
+
+    if origin in (collections.abc.Sequence, list):
+        return check_sequence_or_list(obj, expected_type, path)
+
+    if origin in (dict, typing.Mapping, collections.abc.Mapping) or expected_type in (
+        dict,
+        typing.Mapping,
+        collections.abc.Mapping,
+    ):
+        return check_mapping(obj, expected_type, path)
+
+    if expected_type is int:
+        return check_int(obj, path)
+
+    if expected_type in (float, str, bool):
+        return check_primitive(obj, expected_type, path)
+
+    # Fallback
+    try:
+        if isinstance(obj, expected_type):
+            return TypeCheckResult(True, [])
+        tn = _type_name(expected_type)
+        return TypeCheckResult(False, [f"{path} expected {tn} but got {type(obj).__name__}"])
+    except TypeError:
+        return TypeCheckResult(False, [f"{path} cannot be checked against {expected_type}"])
+
+
+T = TypeVar("T")
+
+
+def ensure_type(obj: object, expected_type: type[T], path: str = "value") -> T:
```

Comment:
`path` is unused?

Reply:
it's passed to `check_type` a few lines down